src/main/java/com/github/manolo8/darkbot/core/objects/swf/FlashMap.java

```diff
@@ -0,0 +1,379 @@
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import static com.github.manolo8.darkbot.Main.API;
+
+/**
+ * Object[K] = V, Array[K] = V, Dictionary[K] = V
+ */
+public class FlashMap<K, V> extends AbstractMap<K, V> implements IUpdatable {
+    /**
+     * since identifiers are always interned strings, they can't be 0,
+     * so we can use 0 as the empty value.
+     */
+    private static final int EMPTY = 0x0;
+
+    // DELETED is stored as the key for deleted items
+    private static final int DELETED = 0x4;
+
+    private static final int kDontEnumBit = 0x01;
+    private static final int kHasDeletedItems = 0x02;
+    private static final int kHasIterIndex = 0x04;
+
+    private static final int MAX_SIZE = 1024;
+    private static final int MAX_CAPACITY = MAX_SIZE * Long.BYTES * 2;
+    private static final byte[] BUFFER = new byte[MAX_CAPACITY];
+
+    private final AtomKind keyKind;
+    private final AtomKind valueKind;
+
+    private final Class<K> keyType;
+    private final Class<V> valueType;
+
+    private final boolean keyUpdatable, valueUpdatable;
+
+    private int size;
+    private Entry[] entries;
+
+    private long address;
+    private Map<K, ValueWrapper> updatables;
+
+    public FlashMap(Class<K> keyType, Class<V> valueType) {
+        this.keyKind = AtomKind.of(keyType);
+        this.valueKind = AtomKind.of(valueType);
+
+        this.keyType = keyType;
+        this.valueType = valueType;
+        if (keyKind.isNotSupported() || valueKind.isNotSupported())
+            throw new IllegalArgumentException("Illegal java types!");
+
+        this.keyUpdatable = Updatable.class.isAssignableFrom(keyType);
+        this.valueUpdatable = Updatable.class.isAssignableFrom(valueType);
+
+        if (keyUpdatable)
+            throw new IllegalArgumentException("Key as updatable is not supported!");
+
+        clear();
+    }
+
+    private static void updateIfChanged(Updatable u, long address) {
+        if (u.address != address)
+            u.update(address);
+    }
+
+    @Override
+    public void update() {
+        if (address == 0) return;
+
+        long traits = API.readLong(address, 16, 40);
+
+        int tableOffset = API.readInt(traits, 236);
+        if (tableOffset == 0) return;
+
+        boolean isDictionary = (API.readInt(traits, 248) & (1 << 4)) != 0;
+        if (isDictionary) {
+            readHashTable(API.readLong(address + tableOffset) + 8); //read hash table ptr & skip cpp vtable
+        } else {
+            readHashTable(address + tableOffset);
+        }
+    }
+
+    @Override
+    public void update(long address) {
+        if (this.address != address) {
+            clear();
+        }
+
+        this.address = address;
+    }
+
+    private int getCapacity(int logCapacity, @SuppressWarnings("unused") boolean hasIterIndexes) {
+        if (logCapacity <= 0) return 0;
+
+        logCapacity = 1 << (logCapacity - 1);
```

Comment:
this looks roughly equivalent to 2 ^ (logCapacity - 1) right? probably should point out in a comment what the more human-readable form of this is